Change Trino to match Hive on column rename/drop/add

djsstarburst opened the pull request:
Issue #6316 shows that if a Hive ORC ACID column is renamed, Hive can't access column data
using the new name, but Trino can.  This is a symptom of a larger problem that Trino and Hive don't
consistently agree after column rename, add or drop.

This commit changes the behavior of Trino to match Hive when columns are renamed,
added or dropped.  Specifically:

* ORC transactional tables were changed to by default find columns by
  column index rather than by name.  ORC non-transactional tables already
  found columns by column index by default.
* Parquet was changed to by default find columns by name rather
  than by column index.
* The Hive config parameter partition_use_column_names, which is used only for ORC
  and Parquet, has been eliminated because it was redundant once the default
  settings for ORC and Parquet were changed to match Hive behavior.

Before this change, if the name of a top-level column in a Hive ORC transactional
table was changed, a subsequent Trino query would not see column data inserted before
the name change, but Hive would see the old data.  After this change, if column names
in old data files differ from column names from the metastore, by default Trino like
Hive will still see the old column values.  This default behavior can be changed by
setting session parameter hive.orc-use-column-names = true.

Similarly, before this change, if the name of a top-level column in a Hive Parquet
table was changed, a subsequent Trino query would see column data inserted before
the name change, but Hive returns null.  After this change, if column names read
from the Parquet data file differ from column names from the metastore, Trino
returns null for renamed column, matching Hive's behavior.  This default behavior
can be changed by setting session parameter hive.parquet-use-column-names = false.

This commit does not address differences between Trino and Hive
in response to column type changes.  This commit also does not address
name or type changes inside complex types like ROW.

The bulk of this commit is a collection of tests that verify Hive
and Trino behavior.

fixes #1558

losipiuk reviewed (commented):
Looks good but I will let someone more familiar with Hive ACID comment.

One general (probably dumb) question. If we can do name correction like that, based on column index, why can't we just use column indexes for identifying columns to queried and not care about potential name mismatches?

djsstarburst commented:
> One general (probably dumb) question. If we can do name correction like that, based on column index, why can't we just use column indexes for identifying columns to queried and not care about potential name mismatches?

Not a dumb question at all, @losipiuk!  I'm a bit nervous using column index alone for the matchup, but like you, I'm uncertain of the guarantees provided by the column index.  I thought about adding another layer of checking - - first matching up by name, and only using the column index if the names disagreed.  We could still do that.  The reason I didn't was I was worried about schema evolution cases where a name was moved from one column to a semantically different column.  I wasn't sure what should happen in that case.  Maybe it's an edge case that I shouldn't be worrying about.

losipiuk commented:
> > One general (probably dumb) question. If we can do name correction like that, based on column index, why can't we just use column indexes for identifying columns to queried and not care about potential name mismatches?
> 
> Not a dumb question at all, @losipiuk! I'm a bit nervous using column index alone for the matchup, but like you, I'm uncertain of the guarantees provided by the column index. I thought about adding another layer of checking - - first matching up by name, and only using the column index if the names disagreed. We could still do that. The reason I didn't was I was worried about schema evolution cases where a name was moved from one column to a semantically different column. I wasn't sure what should happen in that case. Maybe it's an edge case that I shouldn't be worrying about.

Can we verify what Hive does? Maybe it is kinda straightforward from the code (I doubt it :) ). IMO it would be beneficial for the codebase if we could use some immutable identifier for metastore<->ORC column matching. And only fallback to by-name matching if explicitly requested by the user via configuration, as we do for non ACID tables.

findepi commented:
How does this relate to `hive.orc.use-column-names` -- for non-transactional tables, what is Hive's default behavior? What is Hive's default behavior for transactional tables?

How does this relate to non-ORC (insert-only) transactional tables?
We know `hive.parquet.use-column-names` should be true by default (it's currently false by default, and this is a bug), but what about the case when a Parquet table is transactional?

djsstarburst commented:
@electrum and I talked, and based on that discussion and comments from @findepi, it's clear we need more tests.  Here are the tests we think should be written:

- Change the new renames test to run for both transactional and non-transactional tables.
- In the existing renames test, insert records between renames and verify that both old and newly-inserted data is returned by subsequent queries.
- Add a test that demonstrates Hive's behavior in both transactional and non-transaction tables when the last column is dropped, and a new column is added.
- Add a test for transactional and non-transactional Parquet column renames.

djsstarburst commented:
> How does this relate to `hive.orc.use-column-names` -- for non-transactional tables, what is Hive's default behavior? What is Hive's default behavior for transactional tables?

This is spelled out in #6316, however I didn't include all the tests I ran to reach those conclusions.

> How does this relate to non-ORC (insert-only) transactional tables?
> We know `hive.parquet.use-column-names` should be true by default (it's currently false by default, and this is a bug), but what about the case when a Parquet table is transactional?

All good questions.  I hadn't paid attention to `hive.*.use-column-names`.

@electrum and I put together a list of tests that I should write as part of this PR.  I'll figure out how to add use-column-names tests, and tests for insert-only tables.

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveStorageFormat.java:
```diff
@@ -160,6 +165,14 @@ else if (type.getCategory() == Category.PRIMITIVE) {
+
+    public static HiveStorageFormat getHiveStorageFormat(StorageFormat storageFormat)
+    {
+        return requireNonNull(HIVE_STORAGE_FORMAT_FROM_STORAGE_FORMAT.get(storageFormat), "Could not find format for " + storageFormat);
```
Tables can have arbitrary formats, so if the mapping fails, we should assume they are index based.

Since this is for reading, I think we can ignore output format here (just use input format and serde).

djsstarburst replied:
I changed the method return to be Optional<HiveStorageFormat>, and changed the map key to be a class with just the serDe and inputFormat.

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java:
```diff
@@ -426,6 +428,41 @@ else if (column.getBaseHiveColumnIndex() < fileColumns.size()) {
+        for (int index = 0; index < columnCount; index++) {
+            OrcColumn orcColumn = fileColumns.get(index);
+            HiveColumnHandle handle = desiredColumnsByNumber.get(index);
+            if (handle != null) {
```
Maybe simplify to
```java
if (handle != null && !orcColumn.getColumnName().equals(handle.getName())) {
    orcColumn = new OrcColumn(...);
}
builder.add(orcColumn);

djsstarburst replied:
Simplified.

electrum commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java:
```diff
@@ -426,6 +428,41 @@ else if (column.getBaseHiveColumnIndex() < fileColumns.size()) {
+                .collect(toImmutableMap(HiveColumnHandle::getBaseHiveColumnIndex, identity()));
+
+        for (int index = 0; index < columnCount; index++) {
+            OrcColumn orcColumn = fileColumns.get(index);
```
Could shorten the code a bit by naming this `column`

djsstarburst replied:
Shortened.

electrum commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -7616,6 +7617,59 @@ public void testSelectFromPrestoViewReferencingHiveTableWithTimestamps()
+    {
+        String lcFormat = format.name().toLowerCase(Locale.ROOT);
+        Session.SessionBuilder builder = Session.builder(getQueryRunner().getDefaultSession())
+                .setIdentity(Identity.forUser("hive")
```
I don't think this `setIdentity()` call is needed. (it seems redundant in all the other tests)

djsstarburst replied:
Removed.

electrum commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -7616,6 +7617,59 @@ public void testSelectFromPrestoViewReferencingHiveTableWithTimestamps()
+
+        assertUpdate(admin, format("ALTER TABLE %s RENAME COLUMN old_name TO new_name", tableName));
+
+        boolean doesntSeeOldData = formatUseColumnNames || NAMED_COLUMN_ONLY_FORMATS.contains(format);
```
It seems clearer to invert this, since a boolean with "not" creates a double negative: `canSeeOldData`

djsstarburst replied:
Inverted.

djsstarburst commented:
@martint, @findepi and @losipiuk, please take a look at these changes to make Hive ORC and Parquet Trino tables conform to Hive's behavior on column rename/add/drop.

findepi commented:
> Parquet was changed to by default find columns by name rather
than by column index.

thank you ❤️ 


Does it mean we should mark the PR as fixing https://github.com/trinodb/trino/issues/1558?

findepi commented:
> Before this change, if the name of a top-level column in a Hive ORC transactional
table was changed, a subsequent Trino query would not see column data inserted before
the name change. After this change, if column names in old data files differ from
column names from the metastore, by default Trino will still see the old column
values. This default behavior can be changed by setting session parameter
hive.orc_use_column_names = false.

Did you mean `true` here?

findepi commented:
> The Hive config parameter partition_use_column_names, which is used only for ORC
and Parquet, has been eliminated because it was redundant once the default
settings for ORC and Parquet were changed to match Hive behavior.

I am not sure i will miss `partition_use_column_names`  param, but how was it obsoleted by ORC/Parquet changes?
Note that `partition_use_column_names`’s logic was applicable also to tables which are neither ORC or Parquet, or even when some partitions are ORC and some Parquet.

losipiuk reviewed (commented):
Thank you! LGTM. Please comment on https://github.com/trinodb/trino/pull/6479#issuecomment-768125366

djsstarburst commented:
> Did you mean true here?

I sure am glad you are a close reviewer, @findepi!  Changed to "true".

> Does it mean we should mark the PR as fixing #1558?

Yes indeed, this PR fixes #1558.

> I am not sure i will miss partition_use_column_names param, but how was it obsoleted by ORC/Parquet changes?

> Note that partition_use_column_names’s logic was applicable also to tables which are neither ORC or Parquet, or even when some partitions are ORC and some Parquet.

I removed it from the `HiveConfig`, and added it to the list of `@DefunctConfigs` in `HiveConfig`.  The method `isPartitionUseColumnNames()` got moved to `HiveSplitManager`, with this definition:

```
    private static boolean isPartitionUsesColumnNames(ConnectorSession session, HiveStorageFormat storageFormat)
    {
        switch (storageFormat) {
            case AVRO:
                return true;
            case JSON:
                return true;
            case ORC:
                return isUseOrcColumnNames(session);
            case PARQUET:
                return isUseParquetColumnNames(session);
            default:
                return false;
        }
    }
```

Does that do the job, or are there other uses of` isPartitionUseColumnNames/partition_use_column_names/hive.partition-use-column-names` that I somehow overlooked?  Thanks!

findepi commented:
> Yes indeed, this PR fixes #1558.

Great!
Feel free to add "fixes #1558" somewhere in the PR description and then the issue will be linked and closed upon PR merge.


> The method isPartitionUseColumnNames() got moved to HiveSplitManager, with this definition: [...]


That's novel to me, but i like the simplicity of this.
Do we have tests for table/partition schema evolution across storage formats?
Like when a column is removed and a different one is added?

(I think Athena does matching by name for table/partition, but I don't know much about Hive's actual behavior.)

djsstarburst commented:
> Do we have tests for table/partition schema evolution across storage formats?
> Like when a column is removed and a different one is added?

This test exercises column renames for all storage formats: https://github.com/trinodb/trino/pull/6479/commits/52cdcb750a25b8e09dfaf356bb187b5b1ba8b666#diff-a89fe5ce9a9dbbeb63ac62862e3848969d5b8a969fb28653010c196cb7516c0dR7620

I'll add column add/column drop to that test.

martint commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -7616,6 +7617,56 @@ public void testSelectFromPrestoViewReferencingHiveTableWithTimestamps()
+    @Test(dataProvider = "legalUseColumnNamesProvider")
+    public void testUseColumnNames(HiveStorageFormat format, boolean formatUseColumnNames)
+    {
+        String lcFormat = format.name().toLowerCase(Locale.ROOT);
```
Avoid abbreviations in variable names

djsstarburst replied:
Changed to lowerCaseFormat.

electrum commented:
This Athena documentation explains how they handle schema updates: https://docs.aws.amazon.com/athena/latest/ug/handling-schema-updates-chapter.html

martint reviewed (commented):
I haven't looked at the code in depth, but the high-level approach looks good to me.

djsstarburst commented:
> I'll add column add/column drop to that test.

I added `TestHiveIntegrationSmokeTest.testUseColumnAddDrop()`, which tests adding and dropping columns for all formats.

djsstarburst commented:
> (I think Athena does matching by name for table/partition, but I don't know much about Hive's actual behavior.)

> This Athena documentation explains how they handle schema updates:
> https://docs.aws.amazon.com/athena/latest/ug/handling-schema-updates-chapter.html

Athena's behavior depends on the storage format.  I believe that the tests I've added confirm that Trino delivers the same results as documented in the matrix in the link above in response to column rename, add/drop for all storage formats.

findepi commented on testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestHiveTransactionalTable.java:
```diff
@@ -779,6 +779,169 @@ public void testInsertOnlyMultipleWriters(boolean bucketed, HiveOrPresto inserte
+    }
+
+    @Test(groups = HIVE_TRANSACTIONAL)
+    public void testParquetColumnDropAdd()
```
These tests have nothing to do with transactional tables, so please move them out from TestHiveTransactionalTable
